File: frontend/src/lib/api/revenue-targets.ts
/**
 * Monthly revenue targets and the month-pace report.
 *
 * Targets are addressed by **month**, not by id: `period_month` is any date
 * inside the month and the backend normalizes it to the 1st, so `PUT` is a true
 * idempotent upsert and re-saving a month can never fork it into two rows.
 *
 * Everything goes through the spec-typed client so paths, params and responses
 * are checked against `_generated.ts` rather than re-declared by hand.
 */
import { apiClient, type Schemas } from "@/lib/api/_client";

/** A stored monthly target as returned by the API. */
export type RevenueTarget = Schemas["RevenueTargetResponse"];
/** The client-settable fields of one month, plus the month it belongs to. */
export type RevenueTargetUpsert = Schemas["RevenueTargetUpsert"];
export type RevenueTargetList = Schemas["RevenueTargetList"];
/** One funnel stage's actual-vs-required counts for the month. */
export type PaceStage = Schemas["PaceStage"];
export type PaceStageName = PaceStage["stage"];
/** Whether a month is on pace to hit its revenue goal. */
export type RevenuePace = Schemas["RevenuePace"];

/**
 * The planning fields of a target, without the month or the server-owned
 * columns. This is the unit the settings editor copies between a default plan
 * and a per-month override.
 */
export type RevenueTargetPlan = Omit<
  RevenueTargetUpsert,
  "period_month"
>;

export const revenueTargetsApi = {
  /** List a workspace's targets, oldest month first; `year` narrows to one year. */
  list: (workspaceId: string, year?: number): Promise<RevenueTargetList> =>
    apiClient.get("/api/v1/workspaces/{workspace_id}/revenue-targets", {
      path: { workspace_id: workspaceId },
      query: year === undefined ? {} : { year },
    }),

  /**
   * Report whether a month is on pace to hit its revenue goal.
   *
   * A month with no stored target still reports its actuals, flagged with
   * `has_target: false`, so the dashboard can prompt for a goal instead of
   * rendering an error or a wall of zeros.
   */
  getPace: (workspaceId: string, month?: string): Promise<RevenuePace> =>
    apiClient.get("/api/v1/workspaces/{workspace_id}/revenue-targets/pace", {
      path: { workspace_id: workspaceId },
      query: month === undefined ? {} : { month },
    }),

  /** Set (or replace) one month's target. */
  upsert: (
    workspaceId: string,
    target: RevenueTargetUpsert,
  ): Promise<RevenueTarget> =>
    apiClient.put("/api/v1/workspaces/{workspace_id}/revenue-targets", {
      path: { workspace_id: workspaceId },
      body: target,
    }),

  /**
   * Set a season's worth of targets in one atomic statement.
   *
   * All-or-nothing on the backend, so a rejected month never leaves half a year
   * written. The payload may not name the same month twice.
   */
  bulkUpsert: (
    workspaceId: string,
    targets: RevenueTargetUpsert[],
  ): Promise<RevenueTargetList> =>
    apiClient.put("/api/v1/workspaces/{workspace_id}/revenue-targets/bulk", {
      path: { workspace_id: workspaceId },
      body: { targets },
    }),

  /** Clear one month's target. */
  delete: (workspaceId: string, periodMonth: string): Promise<void> =>
    apiClient.del(
      "/api/v1/workspaces/{workspace_id}/revenue-targets/{period_month}",
      { path: { workspace_id: workspaceId, period_month: periodMonth } },
    ),
};

File: backend/app/api/v1/revenue_targets.py
"""Monthly revenue target endpoints, plus the month-pace report.

Thin transport over
:class:`app.services.reporting.revenue_target_service.RevenueTargetService`.
Targets are addressed by **month**, not by id: ``period_month`` is any date
inside the month and the service normalizes it to the 1st, so ``PUT`` is a true
idempotent upsert.

Access follows the neighbouring surfaces, both gates capability-based (the
preferred style for new endpoints — see :mod:`app.api.deps`). Reads require
``reports:view`` like :mod:`app.api.v1.reporting`: a revenue goal and the pace
against it is the same class of information as the P&L. Writes require
``workspace:manage`` because the goal is the owner's commitment, not an
operational setting a dispatcher should be able to move. Both resolve the
caller's workspace membership, and every query underneath is workspace-scoped in
the service.

Fixed paths (``/pace``, ``/bulk``) are declared before ``/{period_month}``:
FastAPI matches in declaration order, and ``pace`` is not a parsable date.
"""

import uuid
from datetime import date
from typing import Annotated

from fastapi import APIRouter, Depends, Path, Query, status

from app.api.deps import (
    DB,
    CanManageWorkspace,
    CanViewReports,
    require_route_capabilities,
)
from app.api.service_errors import ServiceErrorRoute
from app.core.permissions import Capability
from app.schemas.revenue_target import (
    RevenuePace,
    RevenueTargetBulkUpsert,
    RevenueTargetList,
    RevenueTargetResponse,
    RevenueTargetUpsert,
)
from app.services.reporting import RevenueTargetService

router = APIRouter(
    route_class=ServiceErrorRoute,
    dependencies=[
        Depends(require_route_capabilities(Capability.REPORTS_VIEW, Capability.WORKSPACE_MANAGE))
    ],
)

PeriodMonth = Annotated[
    date,
    Path(description="Any date inside the target month; normalized to the 1st"),
]


@router.get("", response_model=RevenueTargetList)
async def list_revenue_targets(
    workspace_id: uuid.UUID,
    membership: CanViewReports,
    db: DB,
    year: Annotated[int | None, Query(ge=2000, le=2100, description="Narrow to one year")] = None,
) -> RevenueTargetList:
    """List a workspace's monthly revenue targets, oldest month first."""
    return await RevenueTargetService(db).list_targets(workspace_id, year=year)


@router.get("/pace", response_model=RevenuePace)
async def get_revenue_pace(
    workspace_id: uuid.UUID,
    membership: CanViewReports,
    db: DB,
    month: Annotated[
        date | None,
        Query(description="Any date inside the month to report (defaults to this month)"),
    ] = None,
) -> RevenuePace:
    """Report whether a month is on pace to hit its revenue goal.

    Months with no stored target still report their actuals, flagged with
    ``has_target: false``.
    """
    return await RevenueTargetService(db).get_pace(workspace_id, month)


@router.put("/bulk", response_model=RevenueTargetList)
async def bulk_upsert_revenue_targets(
    workspace_id: uuid.UUID,
    payload: RevenueTargetBulkUpsert,
    membership: CanManageWorkspace,
    db: DB,
) -> RevenueTargetList:
    """Set a season's worth of targets in one atomic statement."""
    return await RevenueTargetService(db).bulk_upsert(workspace_id, payload)


@router.put("", response_model=RevenueTargetResponse)
async def upsert_revenue_target(
    workspace_id: uuid.UUID,
    payload: RevenueTargetUpsert,
    membership: CanManageWorkspace,
    db: DB,
) -> RevenueTargetResponse:
    """Set (or replace) one month's revenue target."""
    return await RevenueTargetService(db).upsert_target(workspace_id, payload)


@router.get("/{period_month}", response_model=RevenueTargetResponse)
async def get_revenue_target(
    workspace_id: uuid.UUID,
    period_month: PeriodMonth,
    membership: CanViewReports,
    db: DB,
) -> RevenueTargetResponse:
    """Fetch one month's revenue target."""
    return await RevenueTargetService(db).get_target(workspace_id, period_month)


@router.delete("/{period_month}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_revenue_target(
    workspace_id: uuid.UUID,
    period_month: PeriodMonth,
    membership: CanManageWorkspace,
    db: DB,
) -> None:
    """Clear one month's revenue target."""
    await RevenueTargetService(db).delete_target(workspace_id, period_month)
